File: Xin/LeetCode/data_structure/chapterOne/sectionVI.py
"""
剑指 Offer 24. 反转链表
定义一个函数，输入一个链表的头节点，反转该链表并输出反转后链表的头节点。

示例:

输入: 1->2->3->4->5->NULL
输出: 5->4->3->2->1->NULL

限制：

0 <= 节点个数 <= 5000

作者：Krahets
链接：https://leetcode-cn.com/leetbook/read/illustration-of-algorithm/9pdjbm/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

"""


class ListNode:
    def __init__(self, x):
        self.val = x  # 节点值
        self.next = None  # 后继节点引用


# 递归
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        def recur(cur, pre):
            if not cur:
                return pre
            res = recur(cur.next, cur)
            cur.next = pre
            return res

        return recur(head, None)


# 迭代(双指针)
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        cur, pre = head, None
        while cur:
            cur.next, pre, cur = pre, cur, cur.next
        return pre

File: Xin/C++/LeetCode/data_structure/chapterOne/sectionVI.cpp
/*
 * @Author: your name
 * @Date: 2021-02-21 09:43:12
 * @LastEditTime: 2021-02-21 09:47:44
 * @LastEditors: Please set LastEditors
 * @Description: In User Settings Edit
 * @FilePath: \C++\LeetCode\data_structure\chapterOne\sectionVI.cpp
 */

#include <iostream>
using namespace std;

// 迭代
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        ListNode *cur = head, *pre = nullptr;
        while (cur != nullptr)
        {
            ListNode *tmp = cur->next; // 暂存后继节点 cur.next
            cur->next = pre;           // 修改 next 引用指向
            pre = cur;                 // pre 暂存 cur
            cur = tmp;                 // cur 访问下一节点
        }
        return pre;
    }
};

// 递归
class Solution
{
public:
    ListNode *reverseList(ListNode *head)
    {
        return recur(head, nullptr); // 调用递归并返回
    }

private:
    ListNode *recur(ListNode *cur, ListNode *pre)
    {
        if (cur == nullptr)
            return pre;                        // 终止条件
        ListNode *res = recur(cur->next, cur); // 递归后继节点
        cur->next = pre;                       // 修改节点引用指向
        return res;                            // 返回反转链表的头节点
    }
};
